src/parser.rs

```diff
@@ -489,12 +495,34 @@ fn parse_external_id<'a>(pm: &mut XmlMaster<'a>, xml: StringPoint<'a>)
     success(external_id, xml)
 }
 
+fn parse_internal_dtd<'a>(_pm: &mut XmlMaster<'a>, xml: StringPoint<'a>)
+                          -> XmlProgress<'a, &'a str>
+{
+    let (xml, _) = try_parse!(xml.expect_space());
+    let (xml, _) = try_parse!(xml.expect_literal("["));
+    let (xml, _) = xml.consume_space().optional(xml);
+    let (xml, elements) = try_parse!(
+        xml.consume_internal_dtd().map_err(|_| SpecificError::ExpectedInternalDTD)
+    );
+    let (xml, _) = try_parse!(xml.expect_literal("]"));
+
+    success(elements, xml)
+}
+
 /* without the optional intSubset */
 fn parse_document_type_declaration<'a>(pm: &mut XmlMaster<'a>, xml: StringPoint<'a>) -> XmlProgress<'a, Token<'a>> {
     let (xml, _) = try_parse!(xml.expect_literal("<!DOCTYPE"));
     let (xml, _) = try_parse!(xml.expect_space());
     let (xml, _type_name) = try_parse!(xml.consume_name().map_err(|_| SpecificError::ExpectedDocumentTypeName));
-    let (xml, _external_id) = try_parse!(parse_external_id(pm, xml));
```

Comment:
The corresponding grammar here is

```
'<!DOCTYPE' S Name (S ExternalID)? S? ('[' intSubset ']' S?)? '>'
```

This means that 0, 1, or 2 of the external ID and an internal subset can be present. The current code only supports exactly one or the other, closer to this grammar:

```
'<!DOCTYPE' S Name (ext | int) '>'
```

This PR is still an improvement though! I see two paths:

1. If you'd be up for adding more support, then we can add tests for the 0 and 2 cases.
1. If you'd rather stop here, I'd prefer a comment be added to this newly-added `alternate` stating that the grammar isn't yet fully correct.

Reply:
Let me take another quick swing at it. I misunderstood the spec thinking you could have at most one or the other. Interesting.